### DISCUSSION TITLE
Wiring repair at Station

### DISCUSSION THREAD

**Discussion by JamesTCrenshaw** [FakeFishGames/Barotrauma#4743] [Ideas and suggestions]
A common tool of the griefer is to ruin the wiring of the ship, or install something that the rest of the players may not understand how to fix. It seems that most of the time, these problems won't be found until after the game has been saved, and the server now effectively unable to 'roll back' the damages done by the bad actor. 

I propose that, along with the repair hull and items options, that an option exists for the wiring be 'reset' ship-wide to the default placements, while removing any 'new' wiring, like buttons or other placeable wiring, on a needed basis. This would offer a vanilla in-game fix to accidents, griefs, shoddy fix jobs, and other problems associated with so many users messing with the wiring system of the sub. This would also 'punish' the crew for bad wiring by charging the players for having to fix the issues presented.

I would also suggest that such a command could be implemented as a server cheat, to allow fixes on the fly from bad actors, although I am unsure if the game is capable of even doing such a thing.

In either case, this solution would effectively solve a majority of griefing-related resets that occur in multiplayer, and would improve the overall community and multiplayer experience for all players.

**Comment by NilanthAnimosus** [comment]
This is (on a more fundamental level in the code) a pretty difficult thing to accomplish.

We are talking removable/addable components and wires not existing anymore, so needed to add new ones in, or the original ones now in different new places or characters inventories.

This means keeping track of the submarines entire original wiring (Presumably, straight from the sub file even as I cant picture any other conceivable method to pull this off easily). Assuming it can be pulled off, how would the cost be determined before doing it? I suppose one could count existing components and missing ones from the original sub file, then make it an extra 25-75 percent cost repair per component + a base fee. One could get a lot of materials simply undoing the wiring at the end before processing and deconstructing the lot so the cost would have to scale (Though I doubt anyone would be abusing it and caring its abusable if it was).

short of tallying up all components the sub has attached and in inventory separately, then literally deleting all of the attached components then in inventories to cover the difference up, and any remaining bought at extra cost (And oh boy what to do if any logic based attachable pre-existing to the submarine that can contain items could it get more complex that way) ...There's no simple way to balance this in my head or know which component has not moved or changed or been used for other things/crafted with - without major complexity to just add whats missing, and its a pretty complex thing overall to hope for.

If it can actually be done, its a nice interesting idea that would be welcome.

**Reply by mLuby** [reply]
I'm not 100% sure how the other "Repair/Fix all" buttons calculate cost, but they always seems to be around 500mk even when things aren't very broken so my hunch is it's hardcoded. 

Simplest version I can think of would work like this:
1. Captain presses "Reset Wiring and Machinery" (hardcoded cost of 500mk).
2. Game deletes all attached wires _and machines_* from the current sub.
3. Game adds all attached wires and machines and from the original .sub to the current sub.

*It'd have to reset machines too (including switches/lights) because players could have detached and sold lights, to which the game would try and fail to reattach wires.

**Reply by Sviat** [reply]
content of machines needs to be preserved

**Reply by gerakis** [reply]
I would imagine that each wire could have a "mission added from" integer stored in its class. so If you add the wire at the very first station of the game it would be 1, on the first mission it would be 2, on the second station it would be 3, etc. 

Then you could just roll back however far you want with "current mission - x" which would be easy to do because every wire would be tagged with which mission it was first installed at. It turns into a sort of rough timestamp that you can use to figure out what should and shouldnt be there.

**Comment by Kulik242** [comment]
Perhaps, each "wireable" item could have a ghost image of the original wiring (perhaps visible only with high engineering skill) and an option to reset all the values such as, "is on, output, false output, etc." to the original setting? These would probably be not that hard to implement and could help to repair changes even to unskilled players playing engineer. High skill could also add a tag "edited" on the item when inspecting the wiring so it is clear that someone changed something even if the new wiring and underlaying ghost image matches.

**Reply by RadCatish** [reply]
This would help so much!

**Reply by mygamingaccount** [reply]
This is a good idea, but it should be hidden by default, only revealed by something like a checkbox in order to avoid having a cluttered UI

**Comment by Dave247** [comment]
At that point why not just have an option to rest the entire sub to its default/original state? Would be a hell of a lot simpler and easier to implement.
